### PR TITLE
SNOW-1453559: Add support for Series.case_when

### PR DESCRIPTION
**Pull request by sfc-gh-nkumar**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1453559

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Add support for Series.case_when. (does not support callable, follow up JIRA filed SNOW-1489503)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -47,6 +47,7 @@
 - Added support for `Series.str.lstrip` and `Series.str.rstrip`.
 - Added support for `DataFrameGroupby.size` and `SeriesGroupby.size`.
 - Added support for `DataFrame.expanding` and `Series.expanding` for aggregations `count`, `sum`, `min`, `max`, `mean`, `std`, and `var` with `axis=0`.
+- Added support for `Series.case_when` except when condition or replacement is callable.
```
Comment (sfc-gh-nkrishna) [inline CHANGELOG.md:50]: Nit: Can we add this at the end of the new features section?

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 2]
```diff
@@ -7742,6 +7742,125 @@ def getitem_row_array(
...
+        Replace values where the conditions are True.
+
+        Args:
+            caselist: A list of tuples of conditions and expected replacements
+             Takes the form: (condition0, replacement0), (condition1, replacement1), … .
+             condition should be a 1-D boolean array-like object or a callable.
+             replacement should be a 1-D array-like object, a scalar or a callable.
+
+        Returns:
+            New QueryCompiler with replacements.
```
Comment (sfc-gh-nkrishna) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:7756]: Can we add an Examples section to the docstring?

File: tests/integ/modin/series/test_case_when.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    series = native_pd.Series([1, 2, 3, 4, 5, 6])
+    cond1 = native_pd.Series([True, False, True, False, True])
+    cond2 = native_pd.Series([False, True, False, False, True])
+    caselist = [(cond1, 98), (cond2, 99)]
+    _verify_case_when(series, caselist)
+
+
+@pytest.mark.parametrize("caselist", [[], [()], [(97, 98, 99)]])
+@sql_count_checker(query_count=0)
+def test_case_when_invalid_caselist(caselist):
```
Comment (sfc-gh-nkrishna) [inline tests/integ/modin/series/test_case_when.py:60]: Can we also add tests for the TypeErrors raised as well as not implemented for callable?

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 2 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:7756]: Updated docstring in docstrings/series.py

File: CHANGELOG.md (hunk 1 above)
Comment (sfc-gh-nkumar) [reply CHANGELOG.md]: Done

**Review by sfc-gh-nkrishna — APPROVED** [review]
LGTM, thanks Naresh!

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 4]
```diff
@@ -7742,6 +7742,136 @@ def getitem_row_array(
...
+        """
+        # Validate caselist. Errors raises are same as native pandas.
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:7758]:
```suggestion
        # Validate caselist. Errors raised are same as native pandas.
```

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 5]
```diff
@@ -7742,6 +7742,136 @@ def getitem_row_array(
...
+            # against any breaking changes in frontend layer.
+            raise TypeError(
+                f"The caselist argument should be a list; instead got {type(caselist)}"
+            )
+        if not caselist:
+            raise ValueError(
+                "provide at least one boolean condition, with a corresponding replacement."
+            )
+
+        # Validate entries in caselist. Errors raise are same as native pandas.
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:7770]:
```suggestion
        # Validate entries in caselist. Errors raised are same as native pandas.
```

**Review by sfc-gh-helmeleegy — APPROVED** [review]
Looks great. Just a couple of nits. Thanks, Naresh!

**Pull request merged**